File: frontend/src/types/trip.ts
export interface TripRow {
  id: number;
  dispatch_entry_no: string;
  driver_name: string;
  asset_id: string;
  origin_name: string;
  destination_name: string;
  trip_start: string;
  trip_end: string;
  trip_duration_minutes: number;
  eta_met: boolean;
  eta_delay_minutes: number;
  avg_speed_kmph: number;
  trip_km: number;
  trip_status: string;
  is_active: string;
  material_desc: string;
}

export interface TripStats {
  total_trips: number;
  completed_trips: number;
  active_trips: number;
  avg_duration_minutes: number | null;
  avg_speed_kmph: number | null;
  eta_success_rate: number | null;
}

export interface Waypoint {
  latitude: number | null;
  longitude: number | null;
  speed_kmph: number | null;
  status: string | null;
  location_text: string | null;
  distance_from_prev: number | null;
  recorded_at: string;
}

export interface TripDetail {
  trip: TripRow & {
    driver_id: number;
    driver_mobile: string;
    vehicle_id: number;
    asset_type: string;
    origin_id: number;
    destination_id: number;
    customer_id: number;
    customer_name: string;
    trip_eta: string;
    trip_close_remark: string;
  };
  waypoints: Waypoint[];
}

File: backend/app/schemas/trip.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime


class TripOut(BaseModel):
    id: int
    dispatch_entry_no: str
    driver_id: Optional[int] = None
    driver_name: Optional[str] = None
    vehicle_id: Optional[int] = None
    asset_id: Optional[str] = None
    origin_id: Optional[int] = None
    origin_name: Optional[str] = None
    destination_id: Optional[int] = None
    destination_name: Optional[str] = None
    customer_id: Optional[int] = None
    trip_start: Optional[datetime] = None
    trip_end: Optional[datetime] = None
    trip_eta: Optional[datetime] = None
    trip_km: Optional[float] = None
    trip_duration_minutes: Optional[float] = None
    eta_met: Optional[bool] = None
    eta_delay_minutes: Optional[float] = None
    avg_speed_kmph: Optional[float] = None
    trip_status: Optional[str] = None
    is_active: Optional[str] = None
    trip_close_remark: Optional[str] = None
    material_desc: Optional[str] = None


class TripStatsOut(BaseModel):
    total_trips: int
    completed_trips: int
    active_trips: int
    avg_duration_minutes: Optional[float] = None
    avg_speed_kmph: Optional[float] = None
    eta_success_rate: Optional[float] = None
